Disable non-legacy themes selection on Settings screen when AMP slug is defined late

delawski opened the pull request:
## Summary

Fixes #5789

Currently, a theme that cannot be selected as a Reader theme (i.e. defines the AMP slug too late) is disabled in the Onboarding Wizard. However, we don't do the same on the AMP Settings page.

Onboarding Wizard | Settings page
--- | ---
![Onboarding Wizard](https://user-images.githubusercontent.com/134745/105098015-4e2e1280-5a5e-11eb-924d-71b0b575d7bc.png) | ![Settings page](https://user-images.githubusercontent.com/134745/105098086-6b62e100-5a5e-11eb-82a8-80b4f8afb7a4.png)

With this PR, the experience is consistent in both plugin areas:

Onboarding Wizard | Settings page
--- | ---
<img width="814" alt="Screenshot 2021-02-05 at 13 38 35" src="https://user-images.githubusercontent.com/478735/107034785-7d12eb00-67b7-11eb-90c5-c8abd65e046b.png"> | <img width="960" alt="Screenshot 2021-02-05 at 13 38 12" src="https://user-images.githubusercontent.com/478735/107034790-7f754500-67b7-11eb-9800-a13eb1a82e88.png">

Moreover, if there is just one page in the carousel on the Settings page, no navigation will be shown:

Before | After
--- | ---
<img width="969" alt="Screenshot 2021-02-05 at 13 42 18" src="https://user-images.githubusercontent.com/478735/107035156-01fe0480-67b8-11eb-9f80-5a530e912625.png"> | <img width="968" alt="Screenshot 2021-02-05 at 13 41 22" src="https://user-images.githubusercontent.com/478735/107035160-04f8f500-67b8-11eb-8f11-5546eddbd820.png">

## Details

The themes are now categorized by their availability for being used as a Reader Theme in the `ReaderThemes` context provider. We then use the `availableThemes` and `unavailableThemes` lists in both `ReaderThemeCarousel` and `ReaderThemeSelection`.

## Checklist

- [x] My code is tested and passes existing [tests](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines#tests).
- [x] My code follows the [Engineering Guidelines](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines) (updates are often made to the guidelines, check it out periodically).

delawski committed:
Categorize themes by availability in ReaderThemes context provider

A decision whether a theme can be selected as a Reader Theme or not can be done directly in the `ReaderThemes` context provider. We can then use this information both: in the onboarding wizard and on the AMP settings page.

delawski committed:
Remove unused namespace import

delawski committed:
Remove unused namespace imports

delawski committed:
Hide carousel nav if there is just one page

delawski commented:
> Moreover, if there is just one page in the carousel on the Settings page, no navigation will be shown:

@jwold Can you check out the attached screenshots and tell if the update looks good to you?

delawski committed:
Fix indentation issues

delawski committed:
Add unit test to increase coverage

jwold commented:
@delawski thanks for sharing! The screenshots were helpful, makes it easier to get a sense of what's going on.

Here are two small suggestions with the current design:

1. The yellow [notice](https://d.pr/i/qNwx1p) may be a bit strong, instead perhaps a gray notice makes more sense (more in line with the one in the onboarding wizard
2. Instead of a [checkmark](https://d.pr/i/JKRML2) for the `show unavailable themes`, I wonder if we'd want to use a toggle (like we have in other parts of the onboarding wizard. 

And here is a larger suggestion we could look at:

If we are going to have all of the unavailable themes at the end of the list, we could play with a concept to put the notice between the available and unavailable themes. Here's an idea of how that could look:

<img width="512" alt="My Sketches 3 - 2021-02-05 12 31 52" src="https://user-images.githubusercontent.com/3411173/107087498-61f0ad00-67b0-11eb-984c-f81b8689643e.png">

Is that feasible? What do you think?

pierlon commented on assets/src/components/carousel/index.js:
```diff
@@ -124,7 +125,7 @@ export function Carousel( {
 				</ul>
 			</div>
-			{ currentPage && (
+			{ showCarouselNav && (
```
I noticed the carousel navigation is not updated correctly when viewed on mobile:


https://user-images.githubusercontent.com/16200219/107400384-aace9b80-6af9-11eb-8698-8388a3c63977.mp4


A workaround may be to wrap `showCarouselNav` in a `useEffect` hook:

```diff
diff --git a/assets/src/components/carousel/index.js b/assets/src/components/carousel/index.js
--- a/assets/src/components/carousel/index.js	(revision f14aa035c6830468fb1059924477ce6688300a4d)
+++ b/assets/src/components/carousel/index.js	(date 1612887747254)
@@ -38,6 +38,7 @@
 	const { windowWidth } = useWindowWidth();
 	const [ currentPage, originalSetCurrentPage ] = useState( null );
 	const [ pageWidth, setPageWidth ] = useState( 0 );
+	const [ showCarouselNav, setShowCarouselNav ] = useState( null );
 	const carouselListRef = useRef();
 
 	/**
@@ -103,7 +104,13 @@
 		setPageWidth( carouselListRef?.current?.clientWidth || 0 );
 	}, [ items.length, windowWidth ] );
 
-	const showCarouselNav = currentPage && items.length > 1;
+	/**
+	 * Show the carousel navigation when there is more than one page.
+	 */
+	useEffect( () => {
+		setShowCarouselNav( items.length > 1 );
+	}, [ items.length ] );
+
 	const centeredItemIndex = [ ...( carouselListRef.current?.children || [] ) ].indexOf( currentPage );
 	const nextButtonDisabled = centeredItemIndex >= items.length - 1;
 	const prevButtonDisabled = centeredItemIndex <= 0;
```

delawski replied:
Thank you, @pierlon! I've addressed it in 5934dd8.

delawski committed:
Ensure carousel nav state gets updated when items count changes

delawski committed:
Use info notice and toggle for showing unavailable reader themes

delawski commented:
@jwold I've addressed your feedback in f172507. The `info` (light blue) notice type is now used and the checkbox is replaced with a toggle:

<img width="798" alt="Screenshot 2021-02-11 at 19 29 06" src="https://user-images.githubusercontent.com/478735/107681638-9926f880-6c9f-11eb-95f2-464e81ffc05b.png">

I'm wondering if we should update the wording in the message, though. Right now it may be confusing since not *all* themes listed below are unavailable, just some of them.

Maybe it should say something like:

> Some of the themes below are unavailable...

instead of:

> The following themes are unavailable...

jwold commented:
> I'm wondering if we should update the wording in the message, though.

Good catch @delawski . I think that wording change is better.

delawski committed:
Improve the notice microcopy

delawski commented:
> Good catch @delawski . I think that wording change is better.

@jwold This has been updated in 6be58ae

delawski committed:
Merge branch 'develop' into fix/5789-non-legacy-themes-selection

delawski committed:
Merge branch 'develop' into fix/5789-non-legacy-themes-selection

# Conflicts:
#	assets/src/components/reader-theme-selection/index.js
#	assets/src/components/reader-themes-context-provider/index.js
#	src/Admin/OptionsMenu.php

delawski committed:
Pass missing backend props to settings page JS app

pierlon commented on assets/src/components/reader-theme-selection/index.js:
```diff
@@ -53,6 +37,19 @@ export function ReaderThemeSelection() {
+						}
+					</p>
+				</AMPNotice>
+			) }
```
Was this change meant to be apart of the PR? I see it was included in the most recent merge commit: https://github.com/ampproject/amp-wp/pull/5859/commits/4afb475c747cc1ce27b197c88071641d4bbed6b4.

pierlon replied:
Also, should the current theme be disabled here in the Onboarding Wizard? I remember @westonruter reverting the change (see https://github.com/ampproject/amp-wp/pull/5899#discussion_r583289143).

delawski replied:
Thanks, @pierlon. I think this code slipped through my fingers when I've been manually resolving merge conflicts.

delawski committed:
Merge branch 'develop' into fix/5789-non-legacy-themes-selection

delawski committed:
Remove code incorrectly added in merge conflict resolution

delawski committed:
Improve JS test coverage

pierlon committed:
Merge branch 'develop' into fix/5789-non-legacy-themes-selection

pierlon reviewed (approved):
This looks ready to :ship:, great work @delawski!

westonruter reviewed (approved):
Thank you!

jwold commented:
@delawski we had a chat about this on our Tuesday call:

* PG > Haven’t reviewed in a while, will review and - I don’t think there’s much that has changed. Will give my update later today.
* WR > Have idea about this as well, where - reason. On WP.com VIP everything is done in classic VIP in context of a theme; plugins put inside of theme, everything put inside theme. Nothing runs until theme is loaded, and the way you configured the AMP query bar in URL was also done in the theme; you defined AMP query bar constant in theme. If you wanted light or stripped down or ?strippdown, then you put that in your theme. But reader themes need to be running earlier than theme loads, need to switch theme entirely; needs to be done early in WP execution flow, that’s why we disable reader themes when you have defined slug too late. However, in working on the issue with the editor or theme colors, editor color palette; I realized that we could rather store slug in database as well so that even though you defined slug in code, we could essentially cache that so that we can use it earlier in the WP execution cycle even though it’s defined later. 
* WR > Way around this problem
* WR > Number of sites defining slug this late probably low, probably not important to worry about. 
* WR > [As far as next actions] maybe we comment on the issue with the notes, and then think about it for future. 

---
